dev/PowerNotifications/PowerNotifications.cpp

```diff
@@ -197,32 +198,65 @@ namespace winrt::Microsoft::ProjectReunion::implementation
     }
-   
-    // PowerSchemePersonality Functions
-    EventType& PowerSchemePersonality_Event()
+
+    ULONG GetEffectivePowerModeVersion()
+    {
+        // We check for the version supported by checking if the feature is supported
+        // Using NULL as an indicator for uninitialized cache value
+        if (g_powerModeVersion == NULL)
+        {
+            PVOID handle;
+            auto hr = PowerRegisterForEffectivePowerModeNotifications(
+                EFFECTIVE_POWER_MODE_V2, [](EFFECTIVE_POWER_MODE, PVOID) {}, NULL, &handle);
+            g_powerModeVersion = hr == S_OK ? EFFECTIVE_POWER_MODE_V2 : EFFECTIVE_POWER_MODE_V1;
+            check_hresult(PowerUnregisterFromEffectivePowerModeNotifications(handle));
+        }
+        return g_powerModeVersion;
+    }
+
+    EventType& EffectivePowerMode_Event()
     {
-        return make_self<factory_implementation::PowerManager>()->m_powerSchemePersonalityChangedEvent;
+        return make_self<factory_implementation::PowerManager>()->m_powerModeChangedEvent;
     }
 
-    void PowerSchemePersonality_Register()
+    void EffectivePowerMode_Register()
     {
-        check_hresult(RegisterPowerSchemePersonalityChangedListener(
-            &PowerManager::PowerSchemePersonalityChanged_Callback,
-            &make_self<factory_implementation::PowerManager>()->m_powerSchemePersonalityHandle));
+        ULONG version = GetEffectivePowerModeVersion();
+        check_hresult(PowerRegisterForEffectivePowerModeNotifications(
+            version, [](EFFECTIVE_POWER_MODE mode, PVOID)
+            {
+                PowerManager::EffectivePowerModeChanged_Callback(mode);
+            }, nullptr, &g_powerModeHandle));
     }
 
-    void PowerSchemePersonality_Unregister()
+    void EffectivePowerMode_Unregister()
     {
-        check_hresult(UnregisterPowerSchemePersonalityChangedListener(
-            make_self<factory_implementation::PowerManager>()->m_powerSchemePersonalityHandle));
+        check_hresult(PowerUnregisterFromEffectivePowerModeNotifications(g_powerModeHandle));
+        g_powerModeHandle = NULL;
     }
 
-    void PowerSchemePersonality_Update()
+    void EffectivePowerMode_Update()
     {
-        check_hresult(GetPowerSchemePersonality(
-            &make_self<factory_implementation::PowerManager>()->m_cachedPowerSchemePersonality));
+        // Effectively the Get() for PowerMode
+        // Needs to get a temporary subscription to get most recent value
+        // Also, we need a static so that it can be accessed in the lambda below
+        // since references or [&] doesn't work
+        static std::promise<EFFECTIVE_POWER_MODE> promiseObj;
+        static std::future<EFFECTIVE_POWER_MODE> futureObj;
+        PVOID handle;
+        ULONG version = GetEffectivePowerModeVersion();
+        promiseObj = std::promise<EFFECTIVE_POWER_MODE>();
+        futureObj = promiseObj.get_future();
+
+        check_hresult(PowerRegisterForEffectivePowerModeNotifications(
```

Comment:
Don't smuggle out the context in a promise, define a little callback structure with a slim event, like this:

```c++
struct notify_callback {
    EFFECTIVE_POWER_MODE mode;
    wil::slim_event done;
} context;
auto handler = [](EFFECTIVE_POWER_MODE mode, PVOID rawContext) {
    auto context = reinterpret_cast<notify_callback*>(rawContext);
    context->mode = mode;
    context->done.SetEvent();
};
check_hresult(PowerRegisterForEffectivePowerModeNotifications(version, handler, &context, &handle));
context.done.wait();
check_hresult(PowerUnregisterFromEffectivePowerModeNotifications(handle));
```

Reply:
Great idea; changed. Just out of curiosity though, what's wrong with futures/promises?